tests/tests/test_compat.py

```diff
@@ -0,0 +1,295 @@
+    while num_updated < num_devices:
+        if datetime.now(timezone.utc) > deadline:
+            pytest.fail("timeout waiting for devices to submit inventory")
+
+        rsp = api_inventory.call(
+            "GET", inventory.URL_DEVICES, qs_params={"per_page": num_devices * 2},
+        )
+        assert rsp.status_code == 200
+        dev_invs = rsp.json()
+        assert (
+            len(dev_invs) <= num_devices
+        ), "Received more devices from inventory than there exists"
+        if len(dev_invs) < num_devices:
+            time.sleep(1)
+            continue
+        # Check if inventories has been updated since starting this loop
+        num_updated = 0
+        for device in dev_invs:
+            # datetime does not have an RFC3339 parser, but we can convert
+            # the timestamp to a compatible ISO format by removing
+            # fractions and replacing the Zulu timezone with GMT.
+            updated_ts = datetime.fromisoformat(
+                device["updated_ts"].split(".")[0] + "+00:00"
+            )
+            if updated_ts > update_after:
+                num_updated += 1
+            else:
+                time.sleep(1)
+                break
+
+
+def assert_successful_deployment(api_deployments, deployment_id, timeout=TIMEOUT):
+    """
+    Waits for the ongoing deployment (specified by deployment_id) to finish
+    and asserting all devices were successfully upgraded.
+    :param api_deployments: testutils.api.client.ApiClient setup and authorized
+                             to use the deployments management api,
+                             i.e. api_client.with_auth(api_token)
+    :param deployment_id: deployment id to watch
+    :param timeout: optional timeout value to wait for deployment (defaults to 5min)
+    """
+    deadline = datetime.now() + timeout
+    while True:
+        rsp = api_deployments.call(
+            "GET", deployments.URL_DEPLOYMENTS_ID.format(id=deployment_id)
+        )
+        assert rsp.status_code == 200
+
+        dpl = rsp.json()
+        if dpl["status"] == "finished":
+            rsp = api_deployments.call(
+                "GET", deployments.URL_DEPLOYMENTS_STATISTICS.format(id=deployment_id),
+            )
+            assert rsp.status_code == 200
+            assert rsp.json()["failure"] == 0
+            assert rsp.json()["success"] == dpl["device_count"]
+            break
+        elif datetime.now() > deadline:
+            pytest.fail("timeout: Waiting for devices to update")
+        else:
+            time.sleep(1)
+
+
+class TestClientCompatibilityBase:
+    """
+    This class contains compatibility tests implementation for assessing
+    server compatibility with older clients.
+    """
+
+    def compatibility_test_impl(self, env):
+        """
+        The actual test implementation:
+         - Accept devices
+         - Verify devices patches inventory
+         - Perform a noop rootfs update and verify the update was successful
```

Comment:
I think we should do a full rootfs update and check from the clients side that it did reboot. I don't have evidence to point that just exercising the API is not enough, but it just feels so weak as a test. Can you convince me that this is enough?

The added complexity then would be to extract the filesystem from the container to prepare the Artifact (currently we do that for master images in `run.sh` wrapper). Then you could feed it to the Artifact class and still keep using `testutils.api.client.ApiClient` through the test.

The test is very much designed as a backend only test, so I understand that a "noop" update is easier, but if you give up with the we do have helpers in the integration test framework to do the work for us (of course this will

Reply:
I just interpreted the ticket as making a test that make sure that the server never breaks the API for older clients. I can make the client checks more verbose, but it'll come at a cost of significantly increased execution time.

Reply:
I don't have a good argument, really. It might be just enough as-is and it is me being biased by the principle of "an integration test must do stuff on the client".

@kacf Do you have an opinion on this?

Reply:
> I don't have a good argument, really. It might be just enough as-is and it is me being biased by the principle of "an integration test must do stuff on the client".

Well, if I understand it correctly, this *is* doing things on the client, it's just a write optimization. I think the execution time will not be affected that much by doing a real update though. By far the longest time in client tests is the initialization, initial boot, key generation and first login. The rootfs writing is over literally in seconds (this is x86 and HW accelerated), and the reboot you have to do regardless.

@alfrunes is right that it is primarily the API compatibility which is being tested here, so the type of update is not that important. Personally I considered using a `single-file` update, which doesn't require rebooting. I think it boils down to what is easiest and most maintainable in the test framework, so from my PoV a noop update is fine as long as it doesn't complicate things too much. I think my biggest unease would be that it is a bit hacky, but not so much that I would object to doing it! :-)

Reply:
Just to clarify; I know that the test does an update, what I meant is that we don't verify it from the client point of view. For me that is the key difference between integration tests and pure API tests, that in the integration ones we do use the SSH interface to verify the client (what I consider the lower tester).

But then, to use the existing helpers (for the update, or to check the reboot) we would need to extract the file systems from the Docker images of the old clients; which complicates this test framework required for this test. So the noop rootfs update is quite smart, I think.

I am okey with leaving it as is.